src/symbolic/symbolic_choice.ml

```diff
@@ -378,20 +378,29 @@ module Make (Thread : Thread_intf.S) = struct
     in
     f sym
 
+  let stop_if_unreachable pc =
+    let* () = yield in
+    let* solver in
+    match Solver.check solver pc with
+    | `Sat -> modify_thread Thread.mark_pending_pc_done
+    | `Unsat -> stop
+    | `Unknown -> assert false
+
   (*
     Yielding is currently done each time the solver is about to be called,
     in check_reachability and get_model.
   *)
-  let check_reachability v =
-    let* () = yield in
+  let check_reachability =
     let* thread in
-    let* solver in
-    let pc = Thread.pc thread in
-    let sliced_pc = Symbolic_path_condition.slice pc v in
-    match Solver.check solver sliced_pc with
-    | `Sat -> return ()
-    | `Unsat -> stop
-    | `Unknown -> assert false
+    match Thread.pending_pc thread with
+    | [] -> return ()
+    | hd :: tl -> begin
+      let pc_to_check = List.fold_left Symbolic_value.Bool.and_ hd tl in
+      let pc_to_check =
+        Symbolic_path_condition.slice (Thread.pc thread) pc_to_check
```

Comment:
A call to `slice` can only be done if there was a call to `Symbolic_path_condition.add` before.